sn32 rgb: yet another major update - more modularity

dexter93 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

- Decouple key and rgb matrix scanning. Not always required - can live as a standalone rgb driver now
- init bugfixes for `SN32_PWM_OUTPUT_ACTIVE_LOW`
- configuration cleanup - everything is now in a header
- remove most hardcoded values, inherit and/or calculate instead
- driver renamed
- added 26x support
- introducing individual ( per channel) color correction/ luminosity fine tuning

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

dexter93 committed:
sn32f24xb rgb: guard for missing SN32_RGB_MATRIX_ROW_PINS

dexter93 committed:
flesports/fl980: simplify declarations

dexter93 committed:
sn32 shared matrix: autodetect if the matrix is shared

migrate configs in header

dexter93 committed:
sn32 rgb driver: decouple key from rgb matrix

only use custom scanning if the wiring is shared

dexter93 committed:
sn32 rgb driver: make sure init state is off

depending on configuration

dexter93 committed:
sn32 rgb driver: inherit PWM channels from LLD

dexter93 committed:
sn32 rgb driver: Add sn32f260 series support

dexter93 committed:
sn32 rgb driver: prefer `UINT16_MAX` instead of hardcoded value

dexter93 committed:
sn32 rgb driver: fix compilation issue on ROW2COL with COL2ROW PWM direction

dexter93 committed:
sn32 shared matrix: add missing licence header

happy new year!

dexter93 committed:
sn32 rgb driver: add individual channel color correction

LED output differs depending on the module used, as well as the
current limiting setup. Introduce a luminosity correction factor
that will limit output on overdriven channels, allowing fine
tuning through PWM scaling.

dexter93 committed:
drivers: sn32f2xx: simplify name and location

fightforlife commented:
Working good for me on 268F with ROW2COL. But I had two things to change to make it perfect.
Maybe a second board/user would be good to make sure these are no issues on my side/config.


Here I need to use `current_key_col vs last_key_col`, otherwise the keypress effects are lit one to the left.
https://github.com/SonixQMK/qmk_firmware/blob/e3f4c6d48efb4c44aaab522752a6aec3531dfebf/drivers/led/sn32f2xx.c#L518-L522


This is also needed for `HARDWARE_PWM`. If not included, columns which have not the max amount of keys(rows) will light the last key all the time.
https://github.com/SonixQMK/qmk_firmware/blob/e3f4c6d48efb4c44aaab522752a6aec3531dfebf/drivers/led/sn32f2xx.c#L389-L391
https://github.com/SonixQMK/qmk_firmware/blob/e3f4c6d48efb4c44aaab522752a6aec3531dfebf/drivers/led/sn32f2xx.c#L502-L504

dexter93 committed:
sn32 rgb driver: rename `sn32f24xb`-> `sn32f2xx`

dexter93 committed:
sn32: schemas: rename sn32f2xx rgb driver

fightforlife commented on drivers/led/sn32f2xx.c:
```diff
@@ -102,25 +84,28 @@ static uint8_t last_key_col                            = 0;   // key col scan co
 #    endif
 #endif
-#if (DIODE_DIRECTION == ROW2COL)
+#if ((DIODE_DIRECTION == ROW2COL) && (SN32_PWM_DIRECTION != DIODE_DIRECTION))
```
In my ROW2COL board this is not compiling because row_shifter is not definied further down.
When I change it to == it works.

fightforlife commented on drivers/led/sn32f2xx.c:
```diff
@@ -263,13 +379,15 @@ static void update_pwm_channels(PWMDriver *pwmp) {
         uint8_t led_index = g_led_config.matrix_co[current_key_row][current_key_col];
 #    if (SN32_PWM_CONTROL == SOFTWARE_PWM)
-        if (led_index >= SN32F24XB_LED_COUNT) continue;
+        if (led_index >= SN32F2XX_LED_COUNT) continue;
```
For me this line is necessary also for hardware PWM.
If this is not used the last keys in cols which have less then the standard amount of keys.  (e.g. just 4 than normally 5) will light up dimly.

fightforlife commented on drivers/led/sn32f2xx.c:
```diff
@@ -380,7 +500,7 @@ static void update_pwm_channels(PWMDriver *pwmp) {
         uint8_t led_index = g_led_config.matrix_co[current_key_row][current_key_col];
 #    if (SN32_PWM_CONTROL == SOFTWARE_PWM)
-        if (led_index >= SN32F24XB_LED_COUNT) continue;
+        if (led_index >= SN32F2XX_LED_COUNT) continue;
```
For me this line is necessary also for hardware PWM.
If this is not used the last keys in cols which have less then the standard amount of keys.  (e.g. just 4 than normally 5) will light up dimly.

fightforlife commented on drivers/led/sn32f2xx.c:
```diff
@@ -380,7 +500,7 @@ static void update_pwm_channels(PWMDriver *pwmp) {
+        if (led_index >= SN32F2XX_LED_COUNT) continue;
 #    endif
         uint8_t led_row_id = (current_key_row * SN32_RGB_MATRIX_ROW_CHANNELS);
         // Check if we need to enable RGB output
```
In line 519 and 521 I need current_key_col  instead of last_key_col.
Otherwise the keypress effects are shifted one col to the left.

```
    // Enable RGB output
    if (enable_pwm_output) {
#        if (SN32_RGB_OUTPUT_ACTIVE_LEVEL == SN32_RGB_OUTPUT_ACTIVE_HIGH)
        writePinHigh(led_col_pins[current_key_col  ]);
#        elif (SN32_RGB_OUTPUT_ACTIVE_LEVEL == SN32_RGB_OUTPUT_ACTIVE_LOW)
        writePinLow(led_col_pins[current_key_col  ]);
#        endif // SN32_RGB_OUTPUT_ACTIVE_LEVEL
    }
```

fightforlife reviewed (changes requested):
Tested on K66 v1 with ROW2COL hardware PWM shared matrix.
With these 3 changes it is working very good for me.

dexter93 committed:
sn32 rgb driver: update gpio usage

dexter93 committed:
sn32 rgb driver: fix ROW2COL build

dexter93 committed:
sn32 rgb driver: correct SN32_RGB_MATRIX_ROWS_HW init

fix ROW2COL PWM bug on COL2ROW

A column would light up when pressed on this scenario

dexter93 committed:
sn32 rgb driver: cleanup ROW2COL i/o

dexter93 committed:
sn32 rgb driver: rename more keyboards sn32f24xb-> sn32f2xx

dexter93 commented:
@fightforlife after revisiting this and updating ROW2COL scenario, I could not reproduce your issues. If they persist, they are probably related to indexing on your specific board/setup
